### ISSUE TITLE
FPM segfaults with run_on_module

### ISSUE THREAD

**Issue by imbrem** [TheDan64/inkwell#82]
**Describe the Bug**
`fpm.run_on_module(& some module)` with passes segfaults. Runs fine without them though.

**To Reproduce**
```rust
extern crate inkwell;
use inkwell::context::Context;
use inkwell::passes::PassManager;

fn main() {
    let context = Context::create();
    let module = context.create_module("bleh");
    let fpm = PassManager::create_for_function(&module);
    fpm.add_reassociate_pass();
    fpm.initialize();
    fpm.run_on_module(&module);
    println!("Done!");
}
```

**LLVM Version:**
 - LLVM Version: 7.0
 - Inkwell Branch Used: llvm7-0

**Desktop:**
 - OS: Ubuntu 19.04 AND Ubuntu 18.04 (different example tested on the latter in a larger program)

**Comment by TheDan64** [comment]
Isn't `run_on_function` what you want? I wonder if this is because you're creating a function pass manager but running it on a module rather than a function?

Maybe LLVM's naming is just weird but I think you want `PassManager::create_for_module` and then run it on module?

**Comment by imbrem** [comment]
Ah yes that was a mistake on my part. I was modifying Kaleidoscope and decided to run the FPM on the whole module instead of function by function, but I did it wrong. Still, I don't think segfaults should be exposed to user code, right?

Thanks again!

**Comment by TheDan64** [comment]
Yes, this is definitely a bug and should be preventable with subtypes. I started hashing out a solution last night.

**Comment by imbrem** [comment]
Want some help?

**Comment by TheDan64** [comment]
I already started implementing it unfortunately, but I'd certainly be interested in your input on the design:

Make a subtype trait which is implemented for `Module`, `FunctionValue`, and maybe `LTO` (doesn't exist atm). Then make `PassManager` generic over that trait like so:

```rust
pub trait PassManagerSubType {
    type Input;

    unsafe fn populate_from_builder(pass_manager: &PassManager<Self>, builder: &PassManagerBuilder) where Self: Sized;
    unsafe fn create(input: &Self::Input) -> LLVMPassManagerRef;
}

impl PassManagerSubType for Module {
    type Input = ();

    unsafe fn populate_from_builder(pass_manager: &PassManager<Self>, builder: &PassManagerBuilder){
        LLVMPassManagerBuilderPopulateModulePassManager(builder.pass_manager_builder, pass_manager.pass_manager)
    }

    unsafe fn create(_: &()) -> LLVMPassManagerRef {
        LLVMCreatePassManager()
    }
}

// With GATs https://github.com/rust-lang/rust/issues/44265 this could be
// type Input<'a> = &'a Module;
impl PassManagerSubType for FunctionValue {
    type Input = Module;

    unsafe fn populate_from_builder(pass_manager: &PassManager<Self>, builder: &PassManagerBuilder) {
        LLVMPassManagerBuilderPopulateFunctionPassManager(builder.pass_manager_builder, pass_manager.pass_manager)
    }

    unsafe fn create(input: &Module) -> LLVMPassManagerRef {
        LLVMCreateFunctionPassManagerForModule(input.module.get())
    }
}

pub struct PassManager<T> {
    pub(crate) pass_manager: LLVMPassManagerRef,
    sub_type: PhantomData<T>,
}
```

Then you can give it specific impls for T:
```rust
impl<T: PassManagerSubType> PassManager<T> {
    pub fn create(input: &T::Input) -> PassManager<T> {
        let pass_manager = unsafe {
            T::create(input)
        };

        PassManager::new(pass_manager)
    }
}
```

So that users would create and use them like so:
```rust
let mpm = PassManager<Module>::create(());
let fpm = PassManager<FunctionValue>::create(&module);

mpm.run_on(&function); // Compile error
mpm.run_on(&module); // Ok
fpm.run_on(&function); // Ok
fpm.run_on(&module); // Compile error
```

Those compiler errors would effectively prevent those segfaults. And the `PassManagerBuilder` type could have one generic function `populate_pass_manager(pm: PassManager<ST>)` and so on.

Not sure what to do about LTO though, as there is a `LLVMPassManagerBuilderPopulateLTOPassManager` function but it's unclear if it also runs on a specific type like functions or modules only, etc.

**Comment by imbrem** [comment]
Looks good to me, at least so far.

**Comment by TheDan64** [comment]
Fixed in a326bd644885a6d516ae038cd22e24d3604b244b

You should now create your pass managers like so:

```rust
let mpm = PassManager::create(());
let fpm = PassManager::create(&module);

mpm.run_on(&module);
fpm.run_on(&function_value);
```
